### PR TITLE
feat(kork/secrets): Add storage secret engine and s3 implementation

### PR DESCRIPTION
**Pull request by ncknt**
First secret engine implementation

Part of spinnaker/spinnaker#3649

### CODE REVIEW

File: kork-secrets/kork-secrets.gradle [hunk 1]
```diff
@@ -6,6 +6,7 @@ dependencies {
 
   compile spinnaker.dependency('spectatorWebSpring')
   compile spinnaker.dependency('lombok')
+  compile spinnaker.dependency('awsS3')
```
Comment (robzienert) [inline kork-secrets/kork-secrets.gradle]: This ain't gonna fly, unless you want to force any user of this new secrets system to use S3 - nonstarter.

`kork-secrets` should be the API contracts and like... maybe a default file-based (?) implementation or something. Then for S3, you'd wind up making a `kork-secrets-aws` that has an S3 implementation.

File: kork-secrets/src/main/java/com/netflix/spinnaker/config/secrets/engines/AbstractStorageSecretEngine.java [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+import org.springframework.stereotype.Component;
+import org.yaml.snakeyaml.Yaml;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.*;
+
+@Component
```
Comment (robzienert) [inline kork-secrets/src/main/java/com/netflix/spinnaker/config/secrets/engines/AbstractStorageSecretEngine.java]: Why is this a Component? I don't think you can make abstract beans.

File: kork-secrets/src/main/java/com/netflix/spinnaker/config/secrets/engines/AbstractStorageSecretEngine.java [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+public abstract class AbstractStorageSecretEngine implements SecretEngine {
+  protected final static String STORAGE_BUCKET = "b";
+  protected final static String STORAGE_REGION = "r";
+  protected final static String STORAGE_FILE_URI = "f";
+  protected final static String STORAGE_PROP_KEY = "k";
+
+  protected Map<String, Map<String,Object>> cache = new HashMap<>();
+
+  @Autowired
+  protected Yaml yamlParser;
```
Comment (robzienert) [inline kork-secrets/src/main/java/com/netflix/spinnaker/config/secrets/engines/AbstractStorageSecretEngine.java]: Avoid property-based autowiring, always prefer autowired constructors.

File: kork-secrets/src/main/java/com/netflix/spinnaker/config/secrets/engines/AbstractStorageSecretEngine.java [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+
+
+  public String decrypt(EncryptedSecret encryptedSecret) throws SecretDecryptionException {
```
Comment (robzienert) [inline kork-secrets/src/main/java/com/netflix/spinnaker/config/secrets/engines/AbstractStorageSecretEngine.java]: This is a personal preference thing, but I really don't like checked exceptions. Do we need it?

File: kork-secrets/src/main/java/com/netflix/spinnaker/config/secrets/engines/AbstractStorageSecretEngine.java [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+    }
+  }
+
+  public EncryptedSecret encrypt(String secretToEncrypt) throws UnsupportedOperationException {
+    throw new UnsupportedOperationException("This operation is not supported");
+  }
+
+  protected abstract InputStream downloadRemoteFile(EncryptedSecret encryptedSecret) throws IOException;
+
+
```
Comment (robzienert) [inline kork-secrets/src/main/java/com/netflix/spinnaker/config/secrets/engines/AbstractStorageSecretEngine.java]: Mind whitespace.

File: kork-secrets/src/main/java/com/netflix/spinnaker/config/secrets/engines/AbstractStorageSecretEngine.java [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+    }
+  }
+
+  protected void parseAsYaml(String fileURI, InputStream inputStream) {
+    Map<String,Object> parsed = (Map<String, Object>) yamlParser.load(inputStream);
+    cache.put(fileURI, parsed);
+  }
+
+  protected String getParsedValue(String fileURI, String yamlPath) throws SecretDecryptionException {
+    String[] pathElts = yamlPath.split("\\.");
```
Comment (robzienert) [inline kork-secrets/src/main/java/com/netflix/spinnaker/config/secrets/engines/AbstractStorageSecretEngine.java]: You should probably use a `Splitter` instead.

File: kork-secrets/kork-secrets.gradle (hunk 1 above)
Comment (ncknt) [reply kork-secrets/kork-secrets.gradle]: Well, that was the first engine with more to come. The engines register themselves into the registry but the user decides where things are stored. If we move it to a different module, don't we still have the issue of making that jar available at the service startup? Do you want to pick and choose which engines are present when you build your service?

File: kork-secrets/src/main/java/com/netflix/spinnaker/config/secrets/engines/AbstractStorageSecretEngine.java (hunk 2 above)
Comment (ncknt) [reply kork-secrets/src/main/java/com/netflix/spinnaker/config/secrets/engines/AbstractStorageSecretEngine.java]: Oversight and refactor! Fixing...

File: kork-secrets/src/main/java/com/netflix/spinnaker/config/secrets/engines/AbstractStorageSecretEngine.java (hunk 3 above)
Comment (ncknt) [reply kork-secrets/src/main/java/com/netflix/spinnaker/config/secrets/engines/AbstractStorageSecretEngine.java]: This is changing to a simple constructor b/c of the way it's used in the services where we try to instantiate this as early as possible in the Spring lifecycle. So changing too...

File: kork-secrets/src/main/java/com/netflix/spinnaker/config/secrets/engines/AbstractStorageSecretEngine.java (hunk 4 above)
Comment (ncknt) [reply kork-secrets/src/main/java/com/netflix/spinnaker/config/secrets/engines/AbstractStorageSecretEngine.java]: Oversight. That exception should be unchecked. Fixing...

File: kork-secrets/kork-secrets.gradle (hunk 1 above)
Comment (ncknt) [reply kork-secrets/kork-secrets.gradle]: Added `kork-secrets-aws`. Will add the runtime dependency to `kork-core` in another PR

File: kork-secrets/src/main/java/com/netflix/spinnaker/config/secrets/engines/AbstractStorageSecretEngine.java (hunk 4 above)
Comment (ncknt) [reply kork-secrets/src/main/java/com/netflix/spinnaker/config/secrets/engines/AbstractStorageSecretEngine.java]: Fixed

File: kork-secrets/src/main/java/com/netflix/spinnaker/config/secrets/engines/AbstractStorageSecretEngine.java (hunk 6 above)
Comment (ncknt) [reply kork-secrets/src/main/java/com/netflix/spinnaker/config/secrets/engines/AbstractStorageSecretEngine.java]: Added `Splitter`

File: kork-secrets-aws/kork-secrets-aws.gradle [hunk 7]
```diff
@@ -0,0 +1,20 @@
+dependencies {
+  compile spinnaker.dependency('bootAutoConfigure')
+  compile spinnaker.dependency('bootActuator')
+  compile spinnaker.dependency('spectatorGc')
+  compile spinnaker.dependency('spectatorJvm')
+
+  compile spinnaker.dependency('spectatorWebSpring')
+  compile spinnaker.dependency('lombok')
+  compile spinnaker.dependency('awsS3')
```
Comment (robzienert) [inline kork-secrets-aws/kork-secrets-aws.gradle]: I'd just say depend on `kork-core` rather than duplicate all of these.

Comment (ncknt) [reply kork-secrets-aws/kork-secrets-aws.gradle]: Updated

**Pull request merged**